### PR TITLE
BOTMETA Validator + Bulk tidyup, support:(core,community,network)

### PR DESCRIPTION
**Pull request by gundalow**
##### SUMMARY

A lot of PRs were being marked as with GH Label support:core which was
making it difficult for the Network (and other) teams to filter out

* `plugins/` is COMMUNITY
* Set sensible defaults for directories 
* support:network for the platforms that we Networking SUPPORTS,
* everything else is COMMUNITY
* Mark other support:network (ansible-connection, etc)
* Infoblox is support:core
* contrib/ by definition should be support:community
* Remove duplicated labels
* Make yamllint happy(ier)
* Adds **sanity test** to ensure `BOTMETA.yml` is valid

Example output from new sanity test

`ansible-test sanity --python 3.5 --test botmeta`

```
.github/BOTMETA.yml:0:0: extra keys not allowed @ data['files']['lib/ansible/plugins/action/wait_for_connection.py']['foo']. Got 'bar'
.github/BOTMETA.yml:0:0: not a valid value for dictionary value @ data['files']['lib/ansible/plugins/action/win']['support']. Got 'coRe'
```

It's hoped after this has been merged and Ansibullbot has worked over
all the PRs that:
A) labels/support:core will decrease (from 2,907)
B) labels/support:community will increase (from 2,419)
C) labels/support:network will be accurate

Further updates will no doubt be needed, though this should improve the
labels a lot

**Stats as of 2018-09-14**

* pr/issue
* `is:open label:support:core`: 791/2118
* `is:open label:support:network`: 19/114
* `is:open label:support:community`: 991/1,489

Depends on Ansibullbot needs some work:
- [X] `support` propagation https://github.com/ansible/ansibullbot/pull/1010 and 
- [X] RegEx https://github.com/ansible/ansibullbot/pull/1023

##### ISSUE TYPE
 - Bugfix Pull Request
 - Docs Pull Request

### CODE REVIEW

File: .github/BOTMETA.yml
Comment (bcoca) [inline .github/BOTMETA.yml:31]: i would argue it also is used for files WITH it i.e modules

File: .github/BOTMETA.yml
Comment (pilou-) [inline .github/BOTMETA.yml:474]: AFAIK, `support` key [isn't propagated to subpaths](https://github.com/ansible/ansibullbot/blob/f363f697da369fa123d701d74981fdc3d9a52c8e/ansibullbot/parsers/botmetadata.py#L113).

File: .github/BOTMETA.yml
Comment (gundalow) [inline .github/BOTMETA.yml]: @privateip Is `ansible-connection` SUPPORT core or network?

**Comment by gundalow** [comment]
@nitzmahone @ryansb Could you please review this from an AWS, Azure and Windows point of view, in particular:
* Is what I've listed correct in terms of support core/community
* Are their any other plugins I've missed

**Review by ganeshrn — COMMENTED** [review]
Can't find `cli_command` and `cli_config` module entries here

File: .github/BOTMETA.yml
Comment (gundalow) [inline .github/BOTMETA.yml:623]: Add `lib/ansible/plugins/action/cli:`

File: .github/BOTMETA.yml
Comment (gundalow) [reply .github/BOTMETA.yml:31]: I'm not sure if the bot combines support from ANSIBLE_METADATA and BOTMETA.yml.

I don't think we've set `support` for any modules in here.

Comment (pilou-) [reply .github/BOTMETA.yml:31]: > I'm not sure if the bot combines support from ANSIBLE_METADATA and BOTMETA.yml.

Should a file be able to be supported both by `community` and `core` ? (response should be part of the [list of rules being defined](https://github.com/ansible/ansibullbot/pull/1008#pullrequestreview-150104847)) I mean is that `combine` or `override` ?

It seems that Ansibullbot is able to do that - using the `supported_by` keyword, not `support` -  and despite `supported_by` was never used in `BOTMETA.yml`:

1. `supported_by` `BOTMETA` has higher precedence than `supported_by` `ANSIBLE_METADATA` ([see](https://github.com/ansible/ansibullbot/blob/6ab590c39278471986a73d11d13fd6114fdb99f3/ansibullbot/utils/component_tools.py#L218-L219)).
2. when  `supported_by` is set,  `meta[support]` is equal to this variable ([see](https://github.com/ansible/ansibullbot/blob/6ab590c39278471986a73d11d13fd6114fdb99f3/ansibullbot/utils/component_tools.py#L1126))

@gundalow let me know if we should get rid of `supported_by` in `BOTMETA`, meaning:
- either forbidding `BOTMETA` to override `supported_by` from `ANSIBLE_METADATA` 
- or allowing `BOTMETA` to override `supported_by`  from `ANSIBLE_METADATA` using `support` keyword.

Comment (gundalow) [reply .github/BOTMETA.yml:31]: * My understanding is that a file must be supported by exactly ONE entity
* I think `ANSIBLE_METADATA` should override `BOTMETA`. This is because `ANSIBLE_METADATA` is used by the docsite
* AFAIK only Ansible Modules currently have a way of setting `support` via `ANSIBLE_METADATA`, it's possible in the future this may extend to other plugins.

Comment (gundalow) [reply .github/BOTMETA.yml:31]: I wonder if this can be done by ignoring `BOTMETA` `support` for modules, therefore ONLY using what's in `ANSIBLE_METADATA`

File: .github/BOTMETA.yml
Comment (pilou-) [inline .github/BOTMETA.yml]: No need to add `nxos` since it's a path component. When the path is `test/units/modules/network/nxos`, the bot will automatically adds the following labels (if they exists): `test`, `units`, `modules`, `network`, `nxos`.

File: .github/BOTMETA.yml
Comment (pilou-) [inline .github/BOTMETA.yml]: No need to add `nxos` since it's a path component. When the path is `lib/ansible/plugins/terminal/nxos.py`, the bot will automatically adds the following labels (if they exists):  `lib`, `ansible`, `plugins`, `terminal`, `nxos` (`.py` and `.ps1` extensions are [removed](https://github.com/ansible/ansibullbot/blob/149f8e099a8d5822805418c62c73f3fd6256371f/ansibullbot/parsers/botmetadata.py#L79-L80).

File: .github/BOTMETA.yml
Comment (pilou-) [inline .github/BOTMETA.yml]: Should not this label removed too ?

File: .github/BOTMETA.yml
Comment (pilou-) [inline .github/BOTMETA.yml]: s/pluginis/plugins/

File: .github/BOTMETA.yml
Comment (pilou-) [inline .github/BOTMETA.yml:800]: Should not the following entries be added for `connection/oc.py`  ?
```
maintainers: chouseknecht maxamillion fabianvf flaper87 willthames # same as module_utils/k8s
support: community
```

Same for docker connection plugin (maintainers should be all cloud/docker namespace maintainers)

File: .github/BOTMETA.yml
Comment (pilou-) [inline .github/BOTMETA.yml:784]: no need to add `nxos` label, see comments below.

File: .github/BOTMETA.yml
Comment (pilou-) [inline .github/BOTMETA.yml]: s/nxos.ipy/nxos.py/

File: .github/BOTMETA.yml
Comment (pilou-) [inline .github/BOTMETA.yml:716]: No need to use `nxos` label here, see comments below

**Review by ryansb — APPROVED** [review]

File: .github/BOTMETA.yml
Comment (dagwieers) [reply .github/BOTMETA.yml]: Which brings us neatly to the following:
Should we add a hierarchy to integration tests, so they can be auto-labeled too ?

File: .github/BOTMETA.yml
Comment (gundalow) [reply .github/BOTMETA.yml:474]: Fixed by https://github.com/ansible/ansibullbot/pull/1010, thanks Pilou

File: .github/BOTMETA.yml
Comment (gundalow) [reply .github/BOTMETA.yml]: A: Core

File: .github/BOTMETA.yml
Comment (gundalow) [reply .github/BOTMETA.yml:623]: Added

File: .github/BOTMETA.yml
Comment (gundalow) [reply .github/BOTMETA.yml]: Using Ansibullbot to add labels for tests (targets, legacy and units) is tracked in https://github.com/ansible/ansibullbot/issues/808#issuecomment-419853973

File: .github/BOTMETA.yml
Comment (gundalow) [reply .github/BOTMETA.yml]: Good spot, removed.

File: .github/BOTMETA.yml
Comment (gundalow) [reply .github/BOTMETA.yml:800]: Yup. Added
There will no doubt be extra files that are not classified correctly. I'll do some random spot checks on PRs once this has been merged.

**Review by jctanner — APPROVED** [review]

File: .github/BOTMETA.yml
Comment (jctanner) [reply .github/BOTMETA.yml:31]: @gundalow that is a team decision to make. Create a bot issue and add it to a core team agenda or whatever it takes to drive a decision/consensus.

Comment (gundalow) [reply .github/BOTMETA.yml:31]: We currently don't set `support` for anything under `lib/ansible/modules/` in BOTMETA

So I think the discussion here is about future proofing, and possibly adding ANSIBLE_METADATA to other plugin types (which would allow docs to show support status)

So I think this discussion is about future proofing/adding ANSIBLE_METADATA to other plugins

Therefore what's current in BOTMETA and Ansibulbot meta is correct

File: test/sanity/code-smell/botmeta.py
Comment (pilou-) [inline test/sanity/code-smell/botmeta.py]: `voluptuous.MultipleInvalid` should be used instead of `Exception` ([see](https://github.com/alecthomas/voluptuous/blob/e72fd3bfb24f89888f51825313de98c397641674/voluptuous/schema_builder.py#L268-L271) and [voluptuous.error](https://github.com/alecthomas/voluptuous/blob/master/voluptuous/error.py)).

File: test/sanity/code-smell/botmeta.py
Comment (pilou-) [inline test/sanity/code-smell/botmeta.py:16]: Another possible check: `files` keys should match real path.

File: .github/BOTMETA.yml
Comment (pilou-) [inline .github/BOTMETA.yml]: Should not this line (`labels: networking` for `lib/ansible/plugins/netconf/sros.py`) be removed ?

File: .github/BOTMETA.yml
Comment (pilou-) [inline .github/BOTMETA.yml]: one `lib/ansible/plugins/lookup/onepassword` entry could be used instead of `lib/ansible/plugins/lookup/onepassword.py` and `lib/ansible/plugins/lookup/onepassword_raw.py`

File: test/sanity/code-smell/botmeta.py
Comment (gundalow) [reply test/sanity/code-smell/botmeta.py:16]: I guess we'd need to check this matches a regex: https://github.com/ansible/ansibullbot/pull/1023

Is their a case where a file may exist in an older `stable-2.x` branch but not in `devel`, such as when a module is removed/renamed?

Comment (mattclay) [reply test/sanity/code-smell/botmeta.py:16]: That could definitely be the case, particularly when we reorganize files. It makes me wonder if we need per-branch config for the bot -- I'm guessing that would be messy to support and maintain though.

Comment (gundalow) [reply test/sanity/code-smell/botmeta.py:16]: I could see if anything fails the current checks.

I'd prefer to not over engineer this (branched config) at this stage.

Comment (gundalow) [reply test/sanity/code-smell/botmeta.py:16]: I've added path validation and fixed the issues it found, which seem reasonable https://github.com/ansible/ansible/pull/44903/commits/5a899af8466bc91c4791763d9faf9caa952006d0

**Review by pilou- — APPROVED** [review]
LGTM

**Pull request merged**
